include/aws/crt/crypto/SymmetricCipher.h

```diff
@@ -0,0 +1,147 @@
+                 * Returns true on success. Call
+                 * LastError() for the reason this call failed.
+                 */
+                bool Encrypt(const ByteCursor &toEncrypt, ByteBuf &out) noexcept;
+
+                /**
+                 * Encrypts any remaining data on the cipher and stores the output in out. Out can be dynamically
+                 * re-sized if out is a dynamic byte buf. Otherwise, make sure the size of out is at least 2 blocks
+                 * for CBC, CTR, and GCM modes and 40 bytes for KeyWrap.
+                 *
+                 * Returns true on success. Call
+                 * LastError() for the reason this call failed.
+                 */
+                bool FinalizeEncryption(ByteBuf &out) noexcept;
+
+                /**
+                 * Decrypts the value in toEncrypt and stores any immediate results in out. Out can be dynamically
+                 * re-sized if out is a dynamic byte buf. Otherwise, make sure the size of out is at least 1 block
+                 * larger than the input to allow for padding. Returns true on success. Call LastError() for the reason
+                 * this call failed.
+                 */
+                bool Decrypt(const ByteCursor &toDecrypt, ByteBuf &out) noexcept;
+
+                /**
+                 * Decrypts any remaining data on the cipher and stores the output in out. Out can be dynamically
+                 * re-sized if out is a dynamic byte buf. Otherwise, make sure the size of out is at least 2 blocks
+                 * for CBC, CTR, GCM, and Keywrap modes.
+                 *
+                 * Returns true on success. Call
+                 * LastError() for the reason this call failed.
+                 */
+                bool FinalizeDecryption(ByteBuf &out) noexcept;
+
+                bool Reset() noexcept;
+
+                /**
+                 * Returns the key used for this cipher. This key is not copied from the cipher so do not mutate this
+                 * data. Copy if if you need to pass it around anywhere.
+                 */
+                ByteCursor GetKey() const noexcept;
+
+                /**
+                 * Returns the initialization vector used for this cipher.
+                 * This IV is not copied from the cipher so do not mutate this
+                 * data. Copy if if you need to pass it around anywhere.
+                 */
+                ByteCursor GetIV() const noexcept;
+
+                /**
+                 * Returns the encryption tag generated during encryption operations for this cipher in GCM mode.
+                 * This tag is not copied from the cipher so do not mutate this
+                 * data. Copy if if you need to pass it around anywhere.
+                 */
+                ByteCursor GetTag() const noexcept;
+
+              private:
+                SymmetricCipher(aws_symmetric_cipher *cipher) noexcept;
+                aws_symmetric_cipher *m_cipher;

```

Comment:
could we make this a unique ptr instead of raw? we could then in the ctor have something like
```
SymmetricCipher::SymmetricCipher(aws_symmetric_cipher *cipher) noexcept : 
  m_cipher(cipher, aws_symmetric_cipher_destroy), m_lastError(0)
{
  ...
}
```

instead of manually have to call it in the dtor. additonally i think you will be then able to use the default move assignment and ctor as well.